drivers/led/tlc59108.c

```diff
@@ -0,0 +1,212 @@
+		uint32_t delay_on, uint32_t delay_off)
+{
+	struct tlc59108_data *data = dev->data;
+	struct led_data *dev_data = &data->dev_data;
+	uint8_t gdc, gfrq;
+	uint32_t period;
+
+	period = delay_on + delay_off;
+
+	if (period < dev_data->min_period || period > dev_data->max_period) {
+		return -EINVAL;
+	}
+
+	/*
+	 * From manual:
+	 * duty cycle = (GDC / 256) ->
+	 *	(time_on / period) = (GDC / 256) ->
+	 *		GDC = ((time_on * 256) / period)
+	 */
+	gdc = delay_on * 256U / period;
+	if (i2c_reg_write_byte(data->i2c, DT_INST_REG_ADDR(0),
+				TLC59108_GRPPWM,
+				gdc)) {
+		LOG_ERR("LED reg write failed");
+		return -EIO;
+	}
+
+	/*
+	 * From manual:
+	 * period = ((GFRQ + 1) / 24) in seconds.
+	 * So, period (in ms) = (((GFRQ + 1) / 24) * 1000) ->
+	 *		GFRQ = ((period * 24 / 1000) - 1)
+	 */
+	gfrq = (period * 24U / 1000) - 1;
+	if (i2c_reg_write_byte(data->i2c, DT_INST_REG_ADDR(0),
+				TLC59108_GRPFREQ,
+				gfrq)) {
+		LOG_ERR("LED reg write failed");
+		return -EIO;
+	}
+
+	/* Enable blinking mode */
+	if (i2c_reg_update_byte(data->i2c, DT_INST_REG_ADDR(0),
+				TLC59108_MODE2,
+				TLC59108_MODE2_DMBLNK,
+				TLC59108_MODE2_DMBLNK)) {
+		LOG_ERR("LED reg update failed");
+		return -EIO;
+	}
+
+	/* Select the GRPPWM source to drive the LED outpout */
+	return tlc59108_set_ledout(dev, led, TLC59108_LED_GRP_PWM);
+}
+
+static int tlc59108_led_set_brightness(const struct device *dev, uint32_t led,
+		uint8_t value)
+{
+	struct tlc59108_data *data = dev->data;
+	struct led_data *dev_data = &data->dev_data;
+	uint8_t val;
+
+	if (value < dev_data->min_brightness ||
+	    value > dev_data->max_brightness) {
+		return -EINVAL;
+	}
+
+	/* Set the LED brightness value */
+	val = (value * 255U) / dev_data->max_brightness;
+	if (i2c_reg_write_byte(data->i2c, DT_INST_REG_ADDR(0),
+				TLC59108_PWM_BASE + led,
+				val)) {
+		LOG_ERR("LED reg write failed");
+		return -EIO;
+	}
+
+	/* Set the LED driver to be controlled through its PWMx register. */
+	return tlc59108_set_ledout(dev, led, TLC59108_LED_PWM);
+}
+
+static inline int tlc59108_led_on(const struct device *dev, uint32_t led)
+{
+	/* Set LED state to ON */
+	return tlc59108_set_ledout(dev, led, TLC59108_LED_ON);
+}
+
+static inline int tlc59108_led_off(const struct device *dev, uint32_t led)
+{
+	/* Set LED state to OFF */
+	return tlc59108_set_ledout(dev, led, TLC59108_LED_OFF);
+}
+
+static int tlc59108_led_init(const struct device *dev)
+{
+	struct tlc59108_data *data = dev->data;
+	struct led_data *dev_data = &data->dev_data;
+
+	data->i2c = device_get_binding(DT_INST_BUS_LABEL(0));
+	if (data->i2c == NULL) {
+		LOG_DBG("Failed to get I2C device");
+		return -EINVAL;
+	}
+
+	/* Wake up from sleep mode */
+	if (i2c_reg_update_byte(data->i2c, DT_INST_REG_ADDR(0),
+				TLC59108_MODE1,
+				TLC59108_MODE1_OSC, 0)) {
+		LOG_ERR("LED reg update failed");
+		return -EIO;
+	}
+
+	/* Hardware specific limits */
+	dev_data->min_period = 41U;
+	dev_data->max_period = 10667U;
```

Comment:
I see this value as `10.73s` in the datasheet. Any reason for `10.667s`?

Reply:
This driver is pretty much a direct copy of the other driver I'm referring to(NXP PCA9633). That also have 10.73s in the datasheet. I have no idea why that author set that value, but I can't see any good reason for it either. So I will change.